### ISSUE TITLE
Android: Back button not dismissing modals (again)

### ISSUE THREAD

**Issue by infinnie** [driftyco/ionic#9300]
This issue is deliberately a duplicate of #8692.

Reposting here because nobody responsible for this feature has responded to that issue since recently.

**Ionic version:**  (check one with "x")
[ ] **1.x**
[ x ] **2.x**

**I'm submitting a ...**  (check one with "x")
[ x ] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or http://ionicworldwide.herokuapp.com/

**Current behavior:**

I wrote an app with a modal that is expected to be dismissed with either its own “cancel” button or the hardware back button. However since an upgrade to Ionic 2.0.0 RC 1 (RC 2 and 3 as well), it couldn’t be dismissed with the hardware button, but will quit within several presses with the modal present instead. If I manually dismiss the modal after pressing the back button, I could find that the page present is different from the one I opened the modal with, but some previous page.

Nor do hardware back buttons dismiss loading indicators.

**Expected behavior:**

The modal is dismissed with the hardware back button pressed.

**Steps to reproduce:**

1. reproduce and if possible a demo using one of the following templates:
Write an app with a modal
2. Build and run it
3. Open that modal
4. Press the hardware back button and see what would happen.

**Other information:**

**Ionic info:** (run `ionic info` from a terminal/cmd prompt and paste output below):

```
Cordova CLI: 6.3.0
Ionic Framework Version: 2.0.0-rc.3
Ionic CLI Version: 2.1.0
Ionic App Lib Version: 2.1.0-beta.1
OS:
Node Version: v6.3.1
```

Some of the comments: 
> i think it is related with modal not having a ion-nav of its own again. 

> @manucorporat should this be fixed by your modal navigation stuff you working on? @infinnie could you provide a repo we could use to reproduce this issue? Thanks!

> @jgw96 Just build anything that contains a modal for Android.

> Here is one example of register back button, that handles everything:
#6982 (comment)

> @laserus That's not a solution though, that's a workaround. This is definitely a regression bug in the framework.

> @daveshirman yes and no. For modals it is a regression bug, for not-closing-menu-on-back-button it is how driftyco guys see it, for custom alert confirmation on particular pages it is not a bug, but customization (what the backbutton handler is in the first place for).
>
> If you will need to have a custom back button you will end up with similar approach independently of weather there is a bug in modal or not. Unless ionic2 will have some more generic mixing action handling back button (current priority on back button actions does not work for this at all).

> I have the same issue when pushing a page onto the app root. I have a tabs application, and what seems to happen is that after pushing a page onto the root with `this.app.getRootNav().push(NewPage);` pressing the back button moves back through the previously selected tabs (in the background). Only after all the tab history is done, does the new page get dismissed, showing the first tab in the app.

…
> I can confirm that my modals are also not closing via the back-button, on a pretty vanilla (no plugins or JS/css customizations) RC.1 app.
>
> Google Nexus 6. Android 6.0.1

…

> I can confirm this as well. Ionic 2 RC.1 on Android devices.

> @manucorporat any ideas what may be causing this? Is this expected behavior with modals being separate from the main nav?

> @jgw96 Certainly not, if you mean “conforming to the corresponding platform’s recommended behavior for apps” by “expected”. In native Android/WP apps, the hardware back button is intended to be used to dismiss modals if present. Popping up the underlying page without dismissing the modal only makes the whole app confusing.

> Same problem here. RC 2

…

> > Is this expected behavior with modals being separate from the main nav?
>
> This is definitely not expected behaviour, it's really confusing behaviour! 😄 

> Is this ever going to be discussed for a fix? RC2 at the moment, and this behavior is the same since I started developing, back in Beta11.
>
> I personally think it's not supposed to be an expected behavior, it's rather weird. I have tested every app in my phone and used back button on modals and they all dismissed, rather than popping the nav of the page underneath it.

> The system Back button is used to navigate, in reverse chronological order, through the history of screens the user has recently worked with. It is generally based on the temporal relationships between screens, rather than the app's hierarchy.
>
> Android source: https://developer.android.com/design/patterns/navigation.html
>
> If we agree that a modal is a "screen", than it means that the hardware back button should close modals too.
>
> Furthermore, in all the examples listed in the above listed documentation, the hardware back button is never displayed as triggering backs in the background of an opened screen.
>
> So more than just the UIX behavior I may have been excepted, I think that this should be improved to fits the Android guidelines.
>
> Thx in advance for finding an awesome solution as you always do Ionic team ;)

> @manucorporat assign a milestone and fix the problem please. This behavior is annoying and confusing as everybody here reporting the problem knows.

> Not fixed even in RC3…

> @adamdbradley @brandyscarney @manucorporat @danbucholtz @jgw96
>
> Just do something please. If this issue were not solved, I’m afraid that nobody would be willing to use Ionic 2.

> I have to chime in here. Unfortunately this is a serious issue and needs fixing. There is no possible good reason why the back button shouldn't close a modal. It literally makes no sense. This issue is an absolute deal breaker for pushing apps live and needs to be prioritised above style fixes and other component enhancements please.

**Comment by infinnie** [comment]
With the original issue added to the milestone I would close this issue now.
